src/android/shakeDetector.java

```diff
@@ -0,0 +1,251 @@
+
+
+import android.app.Activity;
+import android.app.Service;
+import android.content.Context;
+import android.content.Intent;
+import android.hardware.Sensor;
+import android.hardware.SensorEvent;
+import android.hardware.SensorEventListener;
+import android.hardware.SensorManager;
+import android.os.Handler;
+import android.os.IBinder;
+import android.location.Location;
+
+import com.google.android.gms.location.ActivityRecognitionResult;
+import com.google.android.gms.location.DetectedActivity;
+import com.google.android.gms.location.FusedLocationProviderApi;
+
+public class ShakeDetector extends Service implements SensorEventListener{
+    private SensorManager sensorManager;
+    private Sensor accelerometer;
+    private Sensor gyroscope;
+    private float acceleration;
+    private float currAcceleration;
+    private float lastAcceleration;
+    private long lastUpdated;
+
+    private static final String CONFIG_LIST_KEY = "config_list";
+    private static final String MUTED_LIST_KEY = "muted_list";
+    private static final String POTENTIAL_INCIDENT = "potential_incident";
+    private static final String ID = "id";
+
+    @Override
+    public IBinder onBind(Intent intent){
+        return null;
+    }
+
+    @Override
+    public int onStartCommand(Intent intent, int flags, int startId){
+
+        sensorManager = (SensorManager) getSystemService(Context.SENSOR_SERVICE);
+
+        if(sensorManager.getDefaultSensor(Sensor.TYPE_ACCELEROMETER) != null)
+        {
+            accelerometer = sensorManager.getDefaultSensor(Sensor.TYPE_ACCELEROMETER);
+            sensorManager.registerListener(this, accelerometer, SensorManager.SENSOR_DELAY_NORMAL, new Handler());
+        }
+        else if(sensorManager.getDefaultSensor(Sensor.TYPE_GYROSCOPE) != null)
+        {
+            gyroscope = sensorManager.getDefaultSensor(Sensor.TYPE_GYROSCOPE);
+            sensorManager.registerListener(this, gyroscope, SensorManager.SENSOR_DELAY_NORMAL, new Handler());
+        }
+        return START_STICKY;
+    }
+
+    @Override
+    public void onDestroy() {
+        sensorManager.unregisterListener(this);
+    }
+
+    @Override
+    public  void onAccuracyChanged(Sensor sensor, int accuracy){
+    }
+
+    @Override
+    public void onSensorChanged(SensorEvent event){
+
+        long currTime = System.currentTimeMillis();
+        long deltaTime = currTime - lastUpdated;
+
+
+        //Only get readings if an "accident" has not occured within the past 2 seconds
+        if(deltaTime > 2000) { 
+
+            if (event.sensor.getType() == Sensor.TYPE_ACCELEROMETER) {
+
+                float xAcceleration = event.values[0];
+                float yAcceleration = event.values[1];
+                float zAcceleration = event.values[2];
+
+                //Used to diminish the effect of the acceleration in the y-direction
+                float yAccelerationConst = 0.4f;
+
+                lastAcceleration = currAcceleration;
```

Comment:
This is wrong. What happens if the service is restarted? the data needs to be stored to the usercache and retrieved from there.

Reply:
This only works if the app is running in the foreground or on a non-memory constrained device. And it doesn't store any data to be pushed to the server!